Dev

Airat0694 opened the pull request.

Commit:
delete task

Airat0694 committed:
init dev

AnsGit commented on app/_include/js/script.js:
```diff
@@ -1,4 +1,19 @@
+  // }
+  // alert('sssss');
+
+  function draw(i) {
```
Давай вынесем описание функции в отдельный файл functions.js. И пусть она возвращает экземпляр созданной клетки.

AnsGit replied:
Из полученных клеток нужно сформировать массив, т.к. они еще понадобятся.

AnsGit commented on app/_include/js/script.js:
```diff
@@ -1,4 +1,19 @@
+  function draw(i) {
+    var y = (i < 3) ? 0 : (i < 6) ? 120 : 240;
+    var x = 120 * (i % 3);
+    $("#grid").append('<div class="square square' + i +
```
Я бы оформил так:

```
_div = $("<div>", {"class": "square"})
  .css({"left": x, "top": y})
  .appendTo($("#grid"))
```

AnsGit replied:
И **$("#grid")** нужно сохранить в переменную перед тем, как использовать внутри функций (п.5 в требованиях к ТЗ).

AnsGit commented on app/_include/js/script.js:
```diff
@@ -1,4 +1,19 @@
+  }
+
+  for (var i = 0; i < 9; i++) {
+    draw(i);
```
Давай вынесем фрагменты из HTML и сформируем их по аналогии с сеткой.

AnsGit commented on app/index.html:
```diff
@@ -15,6 +15,24 @@
+          </div>
+
+          <div style="position: relative; top:80px; float:right; width:360px; height:480px">
+            <img style="top:0px; right:0px;" src="_include/img/0_0.png">
```
Давай вынесем фрагменты из HTML и сформируем их по аналогии с сеткой.

AnsGit commented on app/index.html:
```diff
@@ -15,6 +15,24 @@
+          <div id="grid">
+          </div>
+
+          <div style="position: relative; top:80px; float:right; width:360px; height:480px">
```
Стили выносим в css файл (см п.3 в требованиях к ТЗ)

AnsGit commented on app/index.html:
```diff
@@ -15,6 +15,24 @@
+
+          <div style="position: relative; top:80px; float:right; width:360px; height:480px">
+            <img style="top:0px; right:0px;" src="_include/img/0_0.png">
+            <img style="top:120px; right:0px;" src="_include/img/0_1.png">
```
Давай заменим **\<img\>** на **\<div\>**, а изображение будем задавать в виде background.

Airat0694 committed:
Изменение вёрстки:
1) Создан файл function.js, в котором находятся все используемые функции в задании;
2) Сформированы массивы клеток и изображений, присвоены переменным;
3) Изменена функция draw;
4) Некоторые стили вынесены в css файл

Airat0694 committed:
Изменение вёрстки

Airat0694 committed:
изменение вёрстки

AnsGit commented on app/_include/js/functions.js:
```diff
@@ -0,0 +1,31 @@
+    .css({
+      "top": x,
+      "right": y,
+      "background": "url(_include/img/" + z + "_" + (num % 3) + ".png)"
```
Давай создадим 9 классов для 9 фрагментов и к ним в css привяжем background.

Airat0694 committed:
Второй этап: перемешивание фрагментов

AnsGit commented on app/_include/js/script.js:
```diff
@@ -1,4 +1,31 @@
+    cells.push(_cell);
+    _cell.appendTo(grid);
+
+    num = getRandomInt(0, arr.length);
```
Задание состоит несколько в другом. Фрагменты должны справа занимать ту же площадь, что и клетки слева. Но они должны не располагаться просто в ряд, а накладываться друг на друга. Беспорядочно. Т.е. функцию **getRandomInt()** нужно применять по отношению к будущей позиции фрагмента, а не к его порядковому номеру.

Airat0694 committed:
Изменение расположения изображений

Airat0694 committed:
Третий этап: настройка драг фрагментов

AnsGit commented:
1. Если курсор находится за пределами сцены, фрагмент должен продолжать драгаться в пределах нее.
2. Давай будем убивать события **mousemove** и **mouseup**, если событие **mouseup** сработало за пределами сцены. Потому что сейчас, если откликнуть мышь за пределами cцены, а затем вернуть указатель на сцену, фрагмент продолжает драгаться. Подумай, как это можно сделать. И если будут сложности, дай знать.

Airat0694 committed:
Третий этап: настройка драг фрагментов (fix)

AnsGit commented on app/_include/js/script.js:
```diff
@@ -1,4 +1,69 @@
+    pos.parent = parent.offset();
+    pos.cont = cont.offset();
+
+    document.onmousemove = function(event) {
```
Давай все-таки через jQuery сделаем:

```
$(document).on("mousemove", function(event) {
})
```

Тогда везде будет одинаково.

AnsGit commented on app/_include/js/script.js:
```diff
@@ -1,4 +1,69 @@
+      elem.css(new_pos);
+    };
+
+    document.onmouseup = function() {
```
И здесь

Airat0694 committed:
Возвращение изображений на свои места

AnsGit commented on app/_include/js/script.js:
```diff
@@ -1,4 +1,82 @@
+    $(document).on('mouseup', function() {
+
+      $(document).off('mousemove');
+      // console.log();
```
Здесь нужно дополнительно убивать событие **mouseup**:

`$(document).off('mouseup');`

Сейчас получается, что при каждом новом перетаскивании на **document** повторно навешивается событие **mouseup**. Соответственно, при отпускании очередного фрагмента один за другим начинают выполняться обработчики этого события (каждый из которых содержит анимацию до конкретной позиции - позиции предыдущего фрагмента). Поэтому и наблюдается такое поведение.

Airat0694 committed:
Функция проверки постановки фрагмента

AnsGit commented on app/_include/js/script.js:
```diff
@@ -1,4 +1,103 @@
+    cellFull[i] = false;
+  }
+
+  data.cells = cells;
```
Структуру объектов с фрагментами и клетками лучше оформить примерно таким образом:

```
data.cells = [
  {
    self: _cell,
    pos: _cellPos
  },
  ...
]
data.squares = [
  {
    self: _img,
    pos: _imgPos
  },
  ...
]
```

Так будет более наглядно, что к чему.

AnsGit commented on app/_include/js/script.js:
```diff
@@ -1,4 +1,103 @@
+
+    elem = $(this);
+    var pos = {};
+    var num = elem.attr("class").substr(-1);
```
Лучше все-таки сохранить номер в **data()** атрибуте.

Сохранять таким образов: `elem.data({num: 6});`
Получать таким образом: `num = elem.data();`

Тогда не придется делать лишних вычислений с вырезанием номера из класса.

AnsGit replied:
Точно так-же можно сохранить в ячейке информацию о ее заполненности:
`cell.data({full: true})`

AnsGit commented on app/_include/js/functions.js:
```diff
@@ -0,0 +1,130 @@
+    fold = 'yes';
+  }
+
+  if (fold == 'yes') {
```
Необходимо автоматизировать проверку, избавившись от всех условий ниже.

Т.е. нужно следует следующее:
1. Получить текущую позицию фрагмента
2. Пробежаться в цикле по всем ячейкам и зная размеры фрагмента и ячейки проверить, находится ли фрагмент над одной из ячеек.

AnsGit reviewed (changes requested):
1. Сейчас каждую клетку можно занять только 1 раз. Нужно поправить.

Airat0694 committed:
Автоматизация проверки

AnsGit commented on app/_include/js/script.js:
```diff
@@ -1,4 +1,141 @@
+          var cellPos = el.self.position();
+          if ((inCell(pos.new_pos, cellPos)) && (el.self.data(
+              'full') == false)) {
+            if ((data.squares[num].self.data('cellDo') != -1) && (
```
Если подобные длинные выражения встречаются многократно, лучше заранее сохранить их в отдельные переменные:
```
var _cellDo = data.squares[num].self.data('cellDo')
```

AnsGit commented on app/_include/js/script.js:
```diff
@@ -1,4 +1,141 @@
+  var cont = $("#images");
+  var gridPos = $("#grid");
+
+  elems.on('mousedown', function(event) {
```
Если кликнуть на фрагмент, но не двигать его, вываливается ошибка.

AnsGit reviewed (changes requested):
1. Нужно расставить комментарии в ключевых моментах. Код не всегда сходу понятен.

Airat0694 committed:
Добавление комментариев, форматирование кода, исправление ошибки с переносом фрагмента

Airat0694 committed:
Проверка решения

Airat0694 committed:
Проверка решения(fix)

AnsGit commented on app/_include/js/script.js:
```diff
@@ -1,4 +1,186 @@
+      //Проверка решения
+      if (gridFull(data.cells)) {
+
+        $('button').attr('disabled', false).css('background', 'red');
```
Давай вынесем свойство **background** в **css**. А здесь будем оперировать классами:
```
$("button").addClass("wrong")
```

AnsGit replied:
И кнопка должна гореть красным, только если пазл собран с ошибкой.

AnsGit commented on app/_include/js/script.js:
```diff
@@ -1,4 +1,186 @@
+
+        $('button').attr('disabled', false).css('background', 'red');
+
+        $('button').on('click', function() {
```
Это событие нужно навесить на верхнем уровне, - там, где сейчас **mousedown**.

AnsGit replied:
И просто блокировать кнопку в случае, если пазл собран не полностью.

AnsGit reviewed (changes requested):
1. Проверка должна запускаться при нажатии на "Готово", а не автоматически.
2. При ошибке должно появляться сообщение "Неправильно. Попробуй еще раз." (см 3п. в ТЗ: https://docs.google.com/document/d/11g0GCUX7xE9P6wslRVQ-9V15B3JtfJRfF2YTAE1B4g8/edit#heading=h.5k70srqgykg6)

Airat0694 committed:
Написание функции проверки решения

Airat0694 committed:
Написание функции проверки решения(fix)

Airat0694 committed:
Устранение “плавающих” обработчиков событий и анимаций

AnsGit commented on app/_include/js/script.js:
```diff
@@ -1,4 +1,210 @@
+
+  //Проверка постановки всех фрагментов,
+  //Нажатие по кнопке "Готово"
+  $('button').on('click', function() {
```
Нужно все Jquery-объекты (такие, как **$('button')**, **$('.problem')** и другие) сохранить в объекте **data**. И если эти объекты используются внутри функций, передавать их в качестве параметров этих функций, а не обращаться напрямую.

AnsGit replied:
Как, например, в функции **buttonRed()**.

AnsGit commented on app/_include/js/script.js:
```diff
@@ -1,4 +1,210 @@
+  for (var i = 0; i < 9; i++) {
+
+    var _cell = draw(i);
+    var _cellPos = {
```
Позицию можно получить также при помощи Jquery-функции **position()**. Она возвращает объект с ключами **left** и **top**. Код в таком случае сократится до следующего вида:
```
var _cellPos = _cell.position();
```

AnsGit commented on app/_include/js/script.js:
```diff
@@ -1,4 +1,210 @@
+  data.cells = cells;
+  data.squares = squares;
+
+  var elems = $(".img");
```
Эти объекты частично присутствуют в объекте **data**. Нужно все перенести в **data**.

AnsGit commented on app/_include/js/script.js:
```diff
@@ -1,4 +1,210 @@
+    pos.cont = cont.offset();
+    pos.grid = gridPos.offset();
+
+    pos.new_pos = {
```
Нужно написать функцию, которая будет возвращать цельный объект с позицией. Тогда не придется отдельно сохранять **left** и **top**.

AnsGit commented on app/_include/js/script.js:
```diff
@@ -1,4 +1,210 @@
+        left: left,
+        top: top
+      }, 500, function() {
+        if (gridFull(data.cells)) { //Активация кнопки "Готово"
```
В jQuery есть удобная функция **toggleClass()**, которая позволяет добавлять класс при выполнении конкретного условия и удалять его при невыполнении. В данном случае это будет выглядеть так:
```
$('button').toggleClass('disabled', !gridFull(data.cells));
```

AnsGit commented on app/_include/js/script.js:
```diff
@@ -1,4 +1,210 @@
+      $('.scene').fadeOut(2000);
+
+      elems.addClass('disabled');
+      setTimeout(function() {
```
Наверное, нет смысла удалять класс **disabled**, если задача уже завершена.

AnsGit commented on app/_include/js/script.js:
```diff
@@ -1,4 +1,210 @@
+  $('button').on('click', function() {
+
+    if (gridFullTrue(data.squares)) {
+      var actions = [];
```
В данном блоке условия не используется.

AnsGit commented on app/_include/js/script.js:
```diff
@@ -1,4 +1,210 @@
+    if (gridFullTrue(data.squares)) {
+      var actions = [];
+
+      $('.scene').fadeOut(2000);
```
У **fadeOut()**, как и у **fadeIn()**, имеется функция-callback (обратного вызова). Она позволяет выполнить определенные действия после завершения анимации. Нужно использовать ее для того, чтобы убить все события мыши, которые на данный момент существуют. Пример вызова:
```
$('.scene').fadeOut(2000, function() {
  // Действия
});
```

AnsGit commented on app/_include/js/script.js:
```diff
@@ -1,4 +1,210 @@
+            data.cells[_cellId].self.data('full', false);
+            square.self.data('cellInd', -1);
+
+            square.self.animate({
```
Можно и так сделать:
```
square.self.animate(square.pos, 500, function() {});
```

AnsGit replied:
И давай эти анимации занесем в отдельный массив, как на 178 строке. Тогда можно будет точно понять, когда завершатся перемещения всех фрагментов.

AnsGit commented on app/_include/js/script.js:
```diff
@@ -1,4 +1,210 @@
+          }
+        }
+
+        $('.problem').text('Неправильно. Попробуй еще раз.');
```
Пусть сообщение об ошибке появляется после того, как фрагменты вернулись на исходную.

AnsGit replied:
И пусть это сообщение исчезает сразу после правильной расстановки (при нажатии на "Готово").

AnsGit commented on app/_include/js/functions.js:
```diff
@@ -0,0 +1,122 @@
+
+  var left = event.pageX - pos.inner.left -
+    pos.cont.left;
+  var leftMax = pos.parent.left - pos.cont.left + 846;
```
Такие значения, как **6** и **846**, нужно получать в коде. В противном случае при изменении верстки код будет работать некорректно.

Для этих целей могут пригодиться следующие функции:

`pos = elem.position();` - позиция элемента относительно ближайшего родителя с выставленным свойством **position** (**relative** или **absolute**)


`pos = elem.offset();` - позиция элемента относительно верхнего левого угла окна браузера (без большой надобности не использовать).

`prop = elem.css("border");` - получить любое свойство элемента

`prop = elem.attr("width");` - получить любой атрибут элемента.

*(атрибуты - те параметры, которые указываются непосредственно в теге элемента, свойства указываются в атрибуте **style**,  а также в css-файлах)*

`width = elem.width(); height = elem.height();` - возвращают ширину и высоту элемента соответственно.

`width = elem.outerWidth(true); height = elem.outerHeight(true);` - возвращают ширину и высоту элемента соответственно с учетом значения свойств **padding** и **margin** для элемента.

AnsGit reviewed (changes requested):
1. Нужно сделать так, чтобы перетаскиваемый фрагмент располагался поверх всех остальных. Для этого нужно создать отдельный класс с увеличенным **z-index** и навешивать его на перетаскиваемый фрагмент.

2. Справа на сетке при размещении фрагментов остается черный **border**. Нужно избавиться от него.  Можно просто убрать **border** у всех клеток и оставить только общий.

3. Нужно сделать так, чтобы клетка при наведении на нее фрагмента подсвечивалась отдельным цветом. Для этого необходимо создать отдельный класс с измененным цветом и навешивать его на клетку.

4. При наведении мыши на фрагмент курсор должен измениться на **pointer**. При нажатии курсор меняется на **move**. При отпускании курсор сбрасывается.

Курсор при наведении:
```
.img:hover {
  cursor: pointer;
}
```

Airat0694 committed:
Исправление недочётов
